### ISSUE TITLE
Strange issue with et-xmlfile dependency

### ISSUE THREAD

**Issue by Djailla** [pantsbuild/pex#1201]
Hello, since pex > 2.1.24, I'm facing issues with a dependencies of the `tablib` package being an old package et-xmlfile

Here is the log when I run the PEX file 
```± python dist/dcma
Traceback (most recent call last):
  File "/Users/b.vallet/Workspace/python-projects/dcma/dist/dcma/.bootstrap/pex/pex.py", line 476, in execute
  File "/Users/b.vallet/Workspace/python-projects/dcma/dist/dcma/.bootstrap/pex/pex.py", line 132, in activate
  File "/Users/b.vallet/Workspace/python-projects/dcma/dist/dcma/.bootstrap/pex/pex.py", line 119, in _activate
  File "/Users/b.vallet/Workspace/python-projects/dcma/dist/dcma/.bootstrap/pex/environment.py", line 378, in activate
  File "/Users/b.vallet/Workspace/python-projects/dcma/dist/dcma/.bootstrap/pex/environment.py", line 702, in _activate
  File "/Users/b.vallet/Workspace/python-projects/dcma/dist/dcma/.bootstrap/pex/environment.py", line 612, in resolve
pex.environment.ResolveError: Failed to resolve requirements from PEX environment @ /Users/b.vallet/Workspace/python-projects/dcma/dist/dcma.
Needed macosx_10_16_x86_64-cp-36-cp36m compatible dependencies for:
 1: python>=2.6.0
    Required by:
      et-xmlfile 1.0.1
    But this pex had no 'python' distributions.
```

My pex is build with a 'standard' way via tox:

```[testenv:bundle]
deps = pex
setenv =
    LANG=en_US.UTF-8
commands =
    # Collect the statics to be embedded in the sdist and PEX file (via the MANIFEST)
    {envpython} manage.py collectstatic --noinput --clear
    # Creates a source archive in sdist/
    {envpython} setup.py sdist --dist-dir=sdist --format=gztar
    # Build exec file and save it in dist/
    {envpython} setup.py bdist_pex --bdist-dir=dist --pex-args='--disable-cache --index-url=[custom criteo pypi mirror]' --bdist-all
```

Python version is 3.6.8

When building it with pex==2.1.24, I have no issue.

Do you have any clue if I can fix it in my package or the issue comes from the dependency ?

Thanks a lot

**Comment by jsirois** [comment]
Hrm, that package has a bad `setup.py`:
```python
#!/usr/bin/env python

"""Setup script for packaging et_xmfile.

Requires setuptools.

To build the setuptools egg use
    python setup.py bdist_egg
and either upload it to the PyPI with:
    python setup.py upload
or upload to your own server and register the release with PyPI:
    python setup.py register

A source distribution (.zip) can be built with
    python setup.py sdist --format=zip

That uses the manifest.in file for data files rather than searching for
them here.

"""
import codecs
import sys
import os
import warnings
if sys.version_info < (2, 6):
    raise Exception("Python >= 2.6 is required.")
elif sys.version_info[:2] == (3, 2):
    warnings.warn("Python 3.2 is no longer officially supported")

from setuptools import setup, Extension, find_packages
import re

here = os.path.abspath(os.path.dirname(__file__))
try:
    with codecs.open(os.path.join(here, 'README.rst'), encoding="utf-8") as f:
        README = f.read()
except IOError:
    README = ''

from et_xmlfile import (
    __author__,
    __license__,
    __author_email__,
    __url__,
    __version__
)


setup(name='et_xmlfile',
    packages=find_packages(),
    # metadata
    version=__version__,
    description="An implementation of lxml.xmlfile for the standard library",
    long_description=README,
    author=__author__,
    author_email=__author_email__,
    url=__url__,
    license=__license__,
    requires=[
        'python (>=2.6.0)',
        ],
    classifiers=[
                 'Development Status :: 5 - Production/Stable',
                 'Operating System :: MacOS :: MacOS X',
                 'Operating System :: Microsoft :: Windows',
                 'Operating System :: POSIX',
                 'License :: OSI Approved :: MIT License',
                 'Programming Language :: Python',
                 'Programming Language :: Python :: 2.6',
                 'Programming Language :: Python :: 2.7',
                 'Programming Language :: Python :: 3.3',
                 'Programming Language :: Python :: 3.4',
                 ],
    )
```

Note the `requires=['python (>=2.6.0)']`. The `requires` kwarg is old, see here: https://setuptools.readthedocs.io/en/latest/references/keywords.html
```
requires

    Warning

    requires is superseded by install_requires and should not be used anymore.

```

The fact that `requires` is old is not the problem, its that they use that field to restrict the Python version and not for 3rdparty requirements as its intended for. They should be using `python_requires` - in particular `python_requires='>=2.6,!3.2'`.

I'm not really sure what Pex can do here - they're simply using the setup.py metadata incorrectly. It just happens that Pex never used to read this metadata.

You can only use the PEX_IGNORE_ERRORS=1 backdoor at runtime like so:
```
$ pex --version
2.1.28
^jsirois@gill ~/dev/pantsbuild/pex (master) $ pex et-xmlfile==1.0.1 -oet-xmlfile.pex
^jsirois@gill ~/dev/pantsbuild/pex (master) $ ./et-xmlfile.pex 
Traceback (most recent call last):
  File "/home/jsirois/dev/pantsbuild/pex/et-xmlfile.pex/.bootstrap/pex/pex.py", line 476, in execute
  File "/home/jsirois/dev/pantsbuild/pex/et-xmlfile.pex/.bootstrap/pex/pex.py", line 132, in activate
  File "/home/jsirois/dev/pantsbuild/pex/et-xmlfile.pex/.bootstrap/pex/pex.py", line 119, in _activate
  File "/home/jsirois/dev/pantsbuild/pex/et-xmlfile.pex/.bootstrap/pex/environment.py", line 378, in activate
  File "/home/jsirois/dev/pantsbuild/pex/et-xmlfile.pex/.bootstrap/pex/environment.py", line 702, in _activate
  File "/home/jsirois/dev/pantsbuild/pex/et-xmlfile.pex/.bootstrap/pex/environment.py", line 609, in resolve
pex.environment.ResolveError: Failed to resolve requirements from PEX environment @ /home/jsirois/dev/pantsbuild/pex/et-xmlfile.pex.
Needed manylinux_2_32_x86_64-cp-39-cp39 compatible dependencies for:
 1: python>=2.6.0
    Required by:
      et-xmlfile 1.0.1
    But this pex had no 'python' distributions.
^jsirois@gill ~/dev/pantsbuild/pex (master) $ PEX_IGNORE_ERRORS=1 ./et-xmlfile.pex 
Python 3.9.1 (default, Dec 13 2020, 11:55:53) 
[GCC 10.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
(InteractiveConsole)
>>> import et_xmlfile
>>> et_xmlfile.__file__
'/home/jsirois/.pex/installed_wheels/038b58304533e7853e0ae3dc5c617a09f7d1342c/et_xmlfile-1.0.1-py3-none-any.whl/et_xmlfile/__init__.py'
>>> 
now exiting InteractiveConsole...
```

Or else at build-time using by adding `--ignore-errors` to your `--pex-args`.

**Comment by jsirois** [comment]
@Djailla if either the build-time `--ignore-errors` or the runtime `PEX_IGNORE_ERRORS=1` env var solutions work for you, let me know and I'll mark this issue as an answered question.

**Comment by jsirois** [comment]
@Djailla I'm going ahead and closing this as an answered question. Please speak up if that doesn't work for you.

**Comment by jsirois** [comment]
I'm re-opening this since we've had another report of needing et-xmlfile (indirectly), but being unable to use it with modern Pex.

I've done some research in the Pip and setuptools codebase that shows this metadata is acknowledged and even transcribe, but is otherwise ignored. An experiment confirmed this - changing `install_requires` to `requires` in a `setup.py` configuration causes the resulting built distribution, when installed by pip, to not have its dependencies also installed - they are simply skipped!

Afaict from the standards - this is buggy behavior on the part of pip / setuptools:
+ [1.0](https://www.python.org/dev/peps/pep-0241/#fields): No requirement metadata specified yet.
+ [1.1](https://www.python.org/dev/peps/pep-0314/#summary-of-differences-from-pep-241): `Requires` field introduced.
+ [1.2](https://www.python.org/dev/peps/pep-0345/#summary-of-differences-from-pep-314): `Requires-Dist` field introduced and `Requires` field deprecated.
+ [2.1](https://www.python.org/dev/peps/pep-0566/#summary-of-differences-from-pep-345) Says that fields are now specified by the core metadata specification: https://packaging.python.org/specifications/core-metadata/ - that implies `Requires` is no longer valid.

For `et_xmlfile` 1.0.1 which is distributed as an sdist on PyPI we have:
```
$ grep -E "Metadata-Version|Name|Version|Requires" PKG-INFO
Metadata-Version: 1.1
Name: et_xmlfile
Version: 1.0.1
Requires: python (>=2.6.0)
```
So it uses 1.1 metadata wherein `Requires` is valid metadata. Pip and setuptools should interpret it as such and emit more modern metadata if that is their wont, converting Requires to Requires-Dist which 1.2 says is the replacement. 

Since Pex is dealing with the _results_ of `pip {download,wheel, install}` it really can only just go along with this strategy of incorrectly considering `Metadata-Version`. This just means Pex fails to create correct Pexes for the same requirements Pip fails to perform correct resolves for. Without also supporting this same bug during Pexes own validation, it just makes the situation worse for any distribtuon - like `et_xmlfile` that has no real requirements - however you spell them - and so happens not to be broken by Pip / setuptols broken handling (ignoring) of its `Requires` metadata.

**Comment by Djailla** [comment]
Thanks a lot @jsirois will test it as soon as the release is out
